src/main/java/bio/terra/service/snapshot/flight/create/CreateSnapshotPrimaryDataQueryStep.java

```diff
@@ -0,0 +1,92 @@
+package bio.terra.service.snapshot.flight.create;
+
+import bio.terra.grammar.Query;
+import bio.terra.grammar.google.BigQueryVisitor;
+import bio.terra.model.DatasetModel;
+import bio.terra.model.SnapshotRequestContentsModel;
+import bio.terra.model.SnapshotRequestModel;
+import bio.terra.model.SnapshotRequestQueryModel;
+import bio.terra.service.dataset.AssetSpecification;
+import bio.terra.service.dataset.AssetTable;
+import bio.terra.service.dataset.Dataset;
+import bio.terra.service.dataset.DatasetService;
+import bio.terra.service.snapshot.Snapshot;
+import bio.terra.service.snapshot.SnapshotDao;
+import bio.terra.service.tabulardata.google.BigQueryPdao;
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+public class CreateSnapshotPrimaryDataQueryStep implements Step {
+
+    private BigQueryPdao bigQueryPdao;
+    private DatasetService datasetService;
+    private SnapshotDao snapshotDao;
+    private SnapshotRequestModel snapshotReq;
+
+    public CreateSnapshotPrimaryDataQueryStep(BigQueryPdao bigQueryPdao,
+                                              DatasetService datasetService,
+                                              SnapshotDao snapshotDao,
+                                              SnapshotRequestModel snapshotReq) {
+        this.bigQueryPdao = bigQueryPdao;
+        this.datasetService = datasetService;
+        this.snapshotDao = snapshotDao;
+        this.snapshotReq = snapshotReq;
+    }
+
+    @Override
+    public StepResult doStep(FlightContext context) {
+        // TODO: this assumes single-dataset snapshots, will need to add a loop for multiple
+        /*
+         * get dataset and assetName
+         * get asset from dataset
+         * which give the root table
+         * then pass the row id array into create snapshot
+         */
+        SnapshotRequestContentsModel contentsModel = snapshotReq.getContents().get(0);
+        Snapshot snapshot = snapshotDao.retrieveSnapshotByName(snapshotReq.getName());
+        SnapshotRequestQueryModel snapshotQuerySpec = snapshotReq.getContents().get(0).getQuerySpec();
+        String snapshotAssetName = snapshotQuerySpec.getAssetName();
+
+        String snapshotQuery = snapshotReq.getContents().get(0).getQuerySpec().getQuery();
+        Query query = Query.parse(snapshotQuery);
+        List<String> datasetNames = query.getDatasetNames();
```

Comment:
Do we need to validate that what is returned by the query even makes sense (is actually a list of dataset names)? Or is it enough to call `retrieveByName` later?